### PR TITLE
Move cpuid into module

### PR DESCRIPTION
**Pull request by JonathanWoollett-Light**
## Changes

Moves `src/cpuid` into a module in `src/vmm/cpuid/`.

## Reason

By moving the crate into a module code can be more accurately excluded on compilation, raising Arm code coverage in this instance where code is not required on Arm.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md

### CODE REVIEW

File: src/vmm/src/cpuid/common.rs [hunk 1]
```diff
@@ -23,7 +22,7 @@ pub enum GetCpuidError {
 ///
 /// - When the given `leaf` is more than `max_leaf` supported by CPUID.
 /// - When the the CPUID leaf `sub-leaf` is invalid (all its register equal 0).
-#[cfg(cpuid)]
+
```
Comment (kalyazin) [inline src/vmm/src/cpuid/common.rs:25]: are blank lines like this meaningful?

Comment (JonathanWoollett-Light) [reply src/vmm/src/cpuid/common.rs]: No, they are leftover from find/replace I will remove them.

**Review by kalyazin — APPROVED** [review]

File: src/vmm/src/cpuid/cpuid_ffi.rs [hunk 2]
```diff
@@ -532,54 +530,11 @@ impl fmt::LowerHex for RawKvmCpuidEntry {
...
 mod tests {
-    #[cfg(cpuid)]
+
     use kvm_bindings::KVM_MAX_CPUID_ENTRIES;
 
+    use super::super::{CpuidRegisters, CpuidTrait};
     use super::*;
-    use crate::{CpuidRegisters, CpuidTrait};
-
-    #[test]
```
Comment (kalyazin) [inline src/vmm/src/cpuid/cpuid_ffi.rs:541]: are these tests gone now?

File: src/vmm/src/cpuid/intel/normalize.rs [hunk 3]
```diff
@@ -296,211 +297,11 @@ impl super::IntelCpuid {
...
-    #![allow(
-        clippy::undocumented_unsafe_blocks,
-        clippy::unwrap_used,
-        clippy::as_conversions
-    )]
-
-    use super::*;
-
-    #[test]
-    fn default_brand_string_test() {
```
Comment (kalyazin) [inline src/vmm/src/cpuid/intel/normalize.rs:429]: are these gone?

File: src/vmm/src/cpuid/cpuid_ffi.rs (hunk 2 above)
Comment (JonathanWoollett-Light) [reply src/vmm/src/cpuid/cpuid_ffi.rs:541]: I've removed these as they where broken by the change due to the path of the struct changing, they are brittle and I don't believe we  gain much value from testing `Debug` and `Display` implementations

File: src/vmm/src/cpuid/intel/normalize.rs (hunk 3 above)
Comment (JonathanWoollett-Light) [reply src/vmm/src/cpuid/intel/normalize.rs]: No, this was a mistake on my part in the rebase. Fixed.

**Review by zulinx86 — COMMENTED** [review]
Do we still need these `#[cfg(cpuid)]`?
```
$ grep -rn -E "#\[cfg\(cpuid\)\]" src/
src/vmm/src/cpuid/common.rs:167:    #[cfg(cpuid)]
src/vmm/src/cpuid/common.rs:170:    #[cfg(cpuid)]
src/vmm/src/lib.rs:18:#[cfg(cpuid)]
```

File: src/vmm/src/cpuid/intel/normalize.rs [hunk 4]
```diff
@@ -351,7 +352,7 @@ fn default_brand_string(
     const DEFAULT_BRAND_STRING_BASE: &[u8] = b"Intel(R) Xeon(R) Processor @";
 
     // The slice of the host string before the frequency suffix
-    // e.g. b"Intel(R) Xeon(R) Processor Platinum 8275CL CPU @ 3.00" and b"GHz"
+    // e.g. b"Intel(4) Xeon(R) Processor Platinum 8275CL CPU @ 3.00" and b"GHz"
```
Comment (zulinx86) [inline src/vmm/src/cpuid/intel/normalize.rs:355]: should not change.
```suggestion
    // e.g. b"Intel(R) Xeon(R) Processor Platinum 8275CL CPU @ 3.00" and b"GHz"
```

File: src/vmm/src/cpuid/intel/normalize.rs [hunk 5]
```diff
@@ -367,7 +368,7 @@ fn default_brand_string(
     );
 
     // We iterate from the end until hitting a space, getting the frequency number
-    // e.g. b"Intel(R) Xeon(R) Processor Platinum 8275CL CPU @ " and b"3.00"
+    // e.g. b"Intel(4) Xeon(R) Processor Platinum 8275CL CPU @ " and b"3.00"
```
Comment (zulinx86) [inline src/vmm/src/cpuid/intel/normalize.rs:371]: same here
```suggestion
    // e.g. b"Intel(R) Xeon(R) Processor Platinum 8275CL CPU @ " and b"3.00"
```

**Review by zulinx86 — CHANGES_REQUESTED** [review]

File: src/vmm/src/cpuid/intel/mod.rs [hunk 6]
```diff
@@ -67,32 +67,3 @@ impl From<IntelCpuid> for RawCpuid {
...
-        let mut cpuid = IntelCpuid(std::collections::BTreeMap::new());
-        assert_eq!(
-            cpuid.get_mut(&CpuidKey {
-                leaf: 0,
-                subleaf: 0
-            }),
-            None
-        );
-    }
-}
```
Comment (zulinx86) [inline src/vmm/src/cpuid/intel/mod.rs:98]: Why is this part deleted? The tests for AMD still exist.

File: src/cpuid/src/template/amd/mod.rs [hunk 7]
```diff
@@ -1,18 +0,0 @@
...
-use crate::transformer::Error;
-
-pub fn validate_vendor_id() -> Result<(), Error> {
-    let vendor_id = get_vendor_id_from_host()?;
-    if &vendor_id != VENDOR_ID_AMD {
-        return Err(Error::InvalidVendor);
-    }
-
-    Ok(())
-}
```
Comment (zulinx86) [inline src/cpuid/src/template/amd/mod.rs:18]: Do we no longer need this?

File: src/vmm/Cargo.toml [hunk 8]
```diff
@@ -33,8 +33,10 @@ utils = { path = "../utils" }
 virtio_gen = { path = "../virtio_gen" }
 vm-memory = { path = "../vm-memory" }
 
-[target.'cfg(target_arch = "x86_64")'.dependencies]
-cpuid = { path = "../cpuid" }
+arch_gen = { path = "../arch_gen" }
+
+bit-fields = { path = "../bit-fields" }
+log = "0.4.17"
```
[the remark this answers is not in the document]
Comment (JonathanWoollett-Light) [reply src/vmm/Cargo.toml:39]: No these don't need to be separated (moving cpuid to rust-vmm is possible in the future, but not soon).

File: src/vmm/src/cpuid/intel/normalize.rs (hunk 5 above)
Comment (JonathanWoollett-Light) [reply src/vmm/src/cpuid/intel/normalize.rs:371]: Fixed.

File: src/vmm/src/cpuid/intel/mod.rs (hunk 6 above)
Comment (JonathanWoollett-Light) [reply src/vmm/src/cpuid/intel/mod.rs:98]: A mistake in rebase. Fixed.

File: src/cpuid/src/template/amd/mod.rs (hunk 7 above)
Comment (JonathanWoollett-Light) [reply src/cpuid/src/template/amd/mod.rs:18]: Yes, we no longer need these. These have been unneeded since https://github.com/firecracker-microvm/firecracker/pull/3076 but where not removed.

**Review by kalyazin — DISMISSED** [review]

**Comment by JonathanWoollett-Light** [comment]
> Do we still need these `#[cfg(cpuid)]`?
> 
> ```
> $ grep -rn -E "#\[cfg\(cpuid\)\]" src/
> src/vmm/src/cpuid/common.rs:167:    #[cfg(cpuid)]
> src/vmm/src/cpuid/common.rs:170:    #[cfg(cpuid)]
> src/vmm/src/lib.rs:18:#[cfg(cpuid)]
> ```

Only when including the module in `src/vmm`. I will remove this additional occurrences.

**Review by zulinx86 — APPROVED** [review]

**Review by kalyazin — APPROVED** [review]
